utils/s3/client.cc

```diff
@@ -301,42 +302,84 @@ client::group_client& client::find_or_create_client() {
     }
 }
 
-future<> client::make_request(http::request req,
-                              http::experimental::client::reply_handler handle,
-                              std::optional<http::reply::status_type> expected,
-                              seastar::abort_source* as) {
-    auto request = std::move(req);
-    constexpr size_t max_attempts = 3;
-    size_t attempts = 0;
-    while (true) {
-        co_await authorize(request);
-        auto& gc = find_or_create_client();
+static future<> sleep_before_retry(size_t attempted_retries) {
+    if (attempted_retries == 0) {
+        return make_ready_future();
+    }
+    constexpr size_t scale_factor = 25;
+    return seastar::sleep(std::chrono::milliseconds((1UL << attempted_retries) * scale_factor));
+}
+
+http::experimental::client::reply_handler client::wrap_handler(size_t& retry_count,
+                                                               http::request& request,
+                                                               http::experimental::client::reply_handler handler,
+                                                               std::optional<http::reply::status_type> expected) {
+    return [this, &retry_count, &request, expected, handler = std::move(handler)](const http::reply& rep, input_stream<char>&& in) -> future<> {
+        auto _in = std::move(in);
+        auto status_class = seastar::http::reply::classify_status(rep._status);
+        std::optional<aws_error> possible_error;
+        if (status_class != seastar::http::reply::status_class::informational && status_class != seastar::http::reply::status_class::success) {
+            possible_error = aws_error::parse(co_await seastar::util::read_entire_stream_contiguous(_in));
+            if (!possible_error) {
+                possible_error = aws_error::from_http_code(rep._status);
+            }
+        }
+
+        if (possible_error) {
+            auto should_retry = possible_error->is_retryable();
+            if (possible_error->get_error_type() == aws::aws_error_type::REQUEST_TIME_TOO_SKEWED) {
+                s3l.warn("Request failed with REQUEST_TIME_TOO_SKEWED. Machine time: {}, request timestamp: {}",
+                         utils::aws::format_time_point(db_clock::now()),
+                         request.get_header("x-amz-date"));
+                should_retry = aws::retryable::yes;
+            }
+            if (possible_error->get_error_type() == aws::aws_error_type::EXPIRED_TOKEN) {
+                s3l.warn("Request failed with EXPIRED_TOKEN. Resetting credentials");
+                _credentials = {};
+                should_retry = aws::retryable::yes;
+            }
+            co_await authorize(request);
```

Comment:
Why does re-authorization happen on any error?

Reply:
you are right, will fix